libsolidity/parsing/Parser.cpp

```diff
@@ -1614,9 +1614,22 @@ ASTPointer<Expression> Parser::parsePrimaryExpression()
 		}
 		break;
 	case Token::StringLiteral:
+	case Token::HexStringLiteral:
+	{
+		string literal = m_scanner->currentLiteral();
+		Token firstToken = m_scanner->currentToken();
+		while (m_scanner->peekNextToken() == firstToken)
+		{
+			m_scanner->next();
+			literal += m_scanner->currentLiteral();
+		}
 		nodeFactory.markEndPosition();
-		expression = nodeFactory.createNode<Literal>(token, getLiteralAndAdvance());
+		m_scanner->next();
+		if (m_scanner->currentToken() == Token::Illegal)
+			fatalParserError(to_string(m_scanner->currentError()));
```

Comment:
@chriseth I had to add this so that in order to get the error message `Expected even number of hex-nibbles within double-quotes.` instead of `Expected ';' but got 'Illegal'` for [this test](https://github.com/ethereum/solidity/blob/7a2b6dd0e75d164f2764c88fd8f0c1ee2a03c469/test/libsolidity/syntaxTests/string/string_multipart_hex_valid_parts.sol). Is it necessary?

Reply:
I think the solution is to just not call `next()` above in case you have an illegal token.

Reply:
Or maybe you can transform the loop above into a do-while loop?

Reply:
I have tried both of these solution but neither of them worked.

I think that in order for them to work [this line](https://github.com/ethereum/solidity/blob/7a2b6dd0e75d164f2764c88fd8f0c1ee2a03c469/libsolidity/parsing/Parser.cpp#L1674-L1675) which results in the error message `Expected even number of hex-nibbles within double-quotes` should be executed, and in order for this line to be executed, the function [Parser::parsePrimaryExpression](https://github.com/ethereum/solidity/blob/7a2b6dd0e75d164f2764c88fd8f0c1ee2a03c469/libsolidity/parsing/Parser.cpp#L1579) should be called after it parses the hex string but that can't happen, and [this line](https://github.com/ethereum/solidity/blob/7a2b6dd0e75d164f2764c88fd8f0c1ee2a03c469/libsolidity/parsing/Parser.cpp#L1090) is being hit, which is producing the error message `Expected ';' but got 'Illegal'`.

In case I don't call `next()` when an illegal token is found, I'll get the following message instead `Expected ';' but got 'HexStringLiteral'`.

Reply:
Ok - now I understand. I think it's fine like that!